DEV-967 make bib rights date independent

moseshll opened the pull request:
- Fix apparent typo in `config/defaults` referencing `DATAROOT` instead of `DATA_ROOT`.
- Add optional date argument to `run_process_zephir_incremental.sh`.
- Add `run_incremental_zephir.rb` script to calculate a set of dates for `run_process_zephir_incremental.sh`.
- Change error `exit` calls to `exit 1` (may partially or fully address DEV-1003).
- `DBD::mysql` swapped out for `DBD::MariaDB` so GitHub CI can work again.

moseshll committed:
DEV-967 make bib rights date independent
- Add optional date argument to run_process_zephir_incremental.sh
- Add run_incremental_zephir.rb script to calculate a set of dates for run_process_zephir_incremental.sh
- Fix DATA_ROOT vs DATAROOT apparent typo
- Add nonzero result codes to error `exit`s
- Add `PostZephirProcessing::MonthlyInventory` class, integration and unit tests
- `POST_ZEPHIR_LOGGER_LEVEL`

moseshll commented:
@aelkiss This branch might be worth testing out against the Argo workflow since it incorporates a number of fixes.

aelkiss commented on run_incremental_zephir.rb:
```diff
@@ -0,0 +1,47 @@
+
+# Get all of the Zephir files in /htsolr/catalog/prep/ (CATALOG_PREP)
+# of type zephir_upd_YYYYMMDD.json.gz
+all_zephir_upd = Dir.glob(File.join(ENV["CATALOG_PREP"], "zephir_upd_*")).map do |path|
```
`zephir_upd_YYYYMMDD.json.gz` is (one of) the outputs, so we probably need to check which **don't** exist (up to the current date) and handle those... It gets somewhat complicated because we might have downloaded the file from Zephir but not finished processing all the outputs, or we might not have downloaded the file from Zephir at all...

The README has some fairly good documentation on what the files that are inputs and outputs.. but where they end up at various points in the process is less well-documented.

moseshll replied:
The glob part of the new code is for finding what we have; for finding what we still need the logic around line 26 calculates the delta. It would be clearer (not to mention more correct) if the `else` condition were rewritten `dates = ((all_zephir_upd.last + 1)..(Date.today - 1)).to_a.sort`.

Even though this repo isn't ready for a full Ruby-ization, I'll see if I can extract this into `lib/` and add some functionality for looking at some of the other exports like the `.rights` file.

aelkiss replied:
Got it, I think I didn't understand what line 26 was doing.

One case to think about (maybe not for this iteration) is what happens if things run to completion on day N, fail on day N+1, and then work on day N+2. As written, I think this would start running at day N+3 (the day after the last complete update), but I think in reality we would need to go back and process day N+1 followed by day N+2 up through the present (because things need to happen in order...)

moseshll replied:
Looking at a strategy of finding the "earliest gap" where there's one or more days missing from the output directories of interest. And not trying to fill just the gaps (because of deletes) but essentially running back the head and replaying. I'm going on the assumption that the scan for missing dates need not cross month boundaries (e.g. if it is the first of the month on the calendar, the target datestamp is last month so only need to scan thru the files with that month in the datestamp looking for gaps).

aelkiss replied:
That seems right to me.

moseshll replied:
I'm hoping that `PostZephirProcessing::MonthlyInventory` is a better-designed solution. There are some things I would like to do differently once the constraint of using vanilla Ruby is lifted, but I'm reasonably happy with what is here.

aelkiss reviewed (changes requested):
At least as I read this right now (on a Friday afternoon where I am kind of toast) this is backwards - it's looking for the files from dates that were processed successfully and re-trying them?

aelkiss reviewed (commented):
In general this all makes sense; I made a couple of suggestions for refactoring that you can take or leave.

One other thing I was thinking about (especially with respect to testing in kubernetes) is that it might make sense for the script to ensure that all output directories exist (à la the test setup). This was an annoyance I was running into in kubernetes esp. since there are *so many* of them...

moseshll commented on spec/monthly_inventory_test.rb:
```diff
@@ -0,0 +1,138 @@
+def update_file_for_date(date:)
+  File.join(
+    CATALOG_PREP,
+    "zephir_upd_YYYYMMDD.json.gz".sub("YYYYMMDD", date.strftime("%Y%m%d"))
```
I like having that very readable date template -- since there are unhyphenated HT files and hyphenated Zephir files I've tried to highlight difference -- but with the addition of a warning comment I have adopted this succinct variant.

moseshll commented on spec/TODO.txt:
```diff
@@ -0,0 +1,6 @@
+TODO: this is not a real spec since at time of writing we don't have a full modern Ruby
+and Bundler/Gemfile setup with RSpec and the lot. Should not be too hard to adapt once
```
Never would have occurred to me to get the RSpec from Debian. Alas I could not also get `standardrb` but this works, plus I can get coverage.

moseshll commented on lib/monthly_inventory.rb:
```diff
@@ -0,0 +1,118 @@
+      @inventory.each do |_k, v|
+        delta = all_dates - v
+        if delta.count.positive?
+          earliest = earliest.nil? ? delta.min : [earliest, delta.min].min
```
Yes, much neater.

moseshll commented on lib/monthly_inventory.rb:
```diff
@@ -0,0 +1,118 @@
+    # zephir_upd_YYYYMMDD.rights files for the current month.
+    # @return [Array<Date>] sorted ASC
+    def zephir_rights_files
+      dates = directory_inventory(dir: @rights_dir, regexp: RIGHTS_REGEXP)
```
I think my update eliminates some of the DRY violations.

moseshll commented on lib/monthly_inventory.rb:
```diff
@@ -0,0 +1,118 @@
+    # @return [Array<Date>] sorted ASC
+    def directory_inventory(dir:, regexp:)
+      files = []
+      Dir.foreach(dir) do |filename|
```
Ended up using `Dir.children`. I didn't know about `filter_map` or if I did it never occurred to me (Ruby is becoming Lisp!).

aelkiss commented on spec/spec_helper.rb:
```diff
@@ -0,0 +1,184 @@
+
+require_relative "../lib/monthly_inventory"
+
+# This file was generated by the `rspec --init` command. Conventionally, all
```
I'd probably prefer not to have the commented-out boilerplate for rspec & rather have a section that is clearly listed as the 'default options' that we've left enabled vs. the stuff we've added

aelkiss commented on spec/spec_helper.rb:
```diff
@@ -0,0 +1,184 @@
+# avoid stepping on the data/ directory before we can get this new Ruby
+# code better integrated.
+def init_test_tmp
+  Dir.mkdir(TEST_TMP) unless File.directory? TEST_TMP
```
Rather than a static `TEST_TMP` used for all tests, I would rather use a structure like

```ruby
around(:each) do |example|
  Dir.mktmpdir do |tmpdir|
    setup_test_dirs(tmpdir)
    example.run
  end
end
```

This better isolates tests & automatically cleans up the temp directory after each test. (Running under Docker mitigates this somewhat assuming that the tmp directory is somewhere that isn't persistent between container runs).

aelkiss reviewed (commented):
I think this makes more sense to me now. If the tests work reliably as-is (even when interrupted) I think this is fine, but if we run into brittleness I'd like to consider using the `Dir.mktmpdir` approach I mention.

aelkiss commented:
Also just FYI I'd like to try these with argo workflows or at least the test environment in k8s before deploying.

moseshll commented:
@aelkiss I think this should be up to testing vs Argo Workflows at this point. Let me know what you think.

moseshll committed:
Fix missing backslash in deb list

moseshll committed:
DEV-1059 Resolve wide character issue writing to rights_dbm
- UTF-8 encode the TSV written to DBM in `bld_rights_db.pl`
- Add `t/bld_rights_db.t` to call the build script with a crafted value in rights_current.

aelkiss commented:
Same error running in k8s even with the latest commit

```
metadata-processing-test-cwzv8--2655086903: Thu Jan 25 20:00:33 UTC 2024: dump the rights db to a dbm file
metadata-processing-test-cwzv8--2655086903: getting attribute codes
metadata-processing-test-cwzv8--2655086903: getting reason codes
metadata-processing-test-cwzv8--2655086903: prepare
metadata-processing-test-cwzv8--2655086903: execute
metadata-processing-test-cwzv8--2655086903: entering fetch loop
metadata-processing-test-cwzv8--2655086903: Wide character in subroutine entry at /usr/src/app/bld_rights_db.pl line 65.
```
but I will try again after cleaning out any old files that could be potentially interfering... If that doesn't work I can see if I can replicate in a development environment.

aelkiss commented:
Nope, still fails :sob: we can discuss next steps for this

aelkiss commented on run_incremental_zephir.rb:
```diff
@@ -0,0 +1,33 @@
+dates.each do |date|
+  date_str = date.strftime("%Y%m%d")
+  cmd = "#{INCREMENTAL_SCRIPT} #{date_str}"
+  output = `#{cmd}`
```
Using backticks means we don't get the output as it happens, would prefer just using `system` and omitting the `output.length` check below. We should still check the exit status.

moseshll replied:
Just pushed a commit changing backticks to `system`.

aelkiss commented:
See comment above about backticks - once that's fixed I think we can merge this; we can deploy to production and migrate to using the ruby script when we are ready.

moseshll committed:
- Use `system` rather than backticks to execute `run_process_zephir_incremental.sh`.

moseshll committed:
- Replace `docker-compose` with `docker compose` in the GitHub actions, README, and `test_full_file.sh`

aelkiss reviewed (approved):
Looks great. Let's merge it!🚢
